### ISSUE TITLE
[BUG] RuntimeError: cannot pin 'torch.cuda.FloatTensor' only dense CPU tensors can be pinned

### ISSUE THREAD

**Issue by pancho111203** [unit8co/darts#436]
**Describe the bug**
When trying to train the NBEATS model on GPU, I get the error:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-21-cd62ef13c2ee> in <module>
     10     ts_mad,
     11     ts_bcn
---> 12 ], verbose=True)

~/Environments/base/lib/python3.7/site-packages/darts/utils/torch.py in decorator(self, *args, **kwargs)
     63         with fork_rng():
     64             manual_seed(self._random_instance.randint(0, high=MAX_TORCH_SEED_VALUE))
---> 65             return decorated(self, *args, **kwargs)
     66     return decorator

~/Environments/base/lib/python3.7/site-packages/darts/models/torch_forecasting_model.py in fit(self, series, past_covariates, future_covariates, val_series, val_past_covariates, val_future_covariates, verbose, epochs)
    429         logger.info('Train dataset contains {} samples.'.format(len(train_dataset)))
    430 
--> 431         self.fit_from_dataset(train_dataset, val_dataset, verbose, epochs)
    432 
    433     @random_method

~/Environments/base/lib/python3.7/site-packages/darts/utils/torch.py in decorator(self, *args, **kwargs)
     63         with fork_rng():
     64             manual_seed(self._random_instance.randint(0, high=MAX_TORCH_SEED_VALUE))
---> 65             return decorated(self, *args, **kwargs)
     66     return decorator

~/Environments/base/lib/python3.7/site-packages/darts/models/torch_forecasting_model.py in fit_from_dataset(self, train_dataset, val_dataset, verbose, epochs)
    517 
    518         # Train model
--> 519         self._train(train_loader, val_loader, tb_writer, verbose, train_num_epochs)
    520 
    521         # Close tensorboard writer

~/Environments/base/lib/python3.7/site-packages/darts/models/torch_forecasting_model.py in _train(self, train_loader, val_loader, tb_writer, verbose, epochs)
    790             total_loss = 0
    791 
--> 792             for batch_idx, train_batch in enumerate(train_loader):
    793                 self.model.train()
    794                 output = self._produce_train_output(train_batch[:-1])

~/Environments/base/lib/python3.7/site-packages/torch/utils/data/dataloader.py in __next__(self)
    515             if self._sampler_iter is None:
    516                 self._reset()
--> 517             data = self._next_data()
    518             self._num_yielded += 1
    519             if self._dataset_kind == _DatasetKind.Iterable and \

~/Environments/base/lib/python3.7/site-packages/torch/utils/data/dataloader.py in _next_data(self)
    557         data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
    558         if self._pin_memory:
--> 559             data = _utils.pin_memory.pin_memory(data)
    560         return data
    561 

~/Environments/base/lib/python3.7/site-packages/torch/utils/data/_utils/pin_memory.py in pin_memory(data)
     53         return type(data)(*(pin_memory(sample) for sample in data))
     54     elif isinstance(data, container_abcs.Sequence):
---> 55         return [pin_memory(sample) for sample in data]
     56     elif hasattr(data, "pin_memory"):
     57         return data.pin_memory()

~/Environments/base/lib/python3.7/site-packages/torch/utils/data/_utils/pin_memory.py in <listcomp>(.0)
     53         return type(data)(*(pin_memory(sample) for sample in data))
     54     elif isinstance(data, container_abcs.Sequence):
---> 55         return [pin_memory(sample) for sample in data]
     56     elif hasattr(data, "pin_memory"):
     57         return data.pin_memory()

~/Environments/base/lib/python3.7/site-packages/torch/utils/data/_utils/pin_memory.py in pin_memory(data)
     45 def pin_memory(data):
     46     if isinstance(data, torch.Tensor):
---> 47         return data.pin_memory()
     48     elif isinstance(data, string_classes):
     49         return data

RuntimeError: cannot pin 'torch.cuda.FloatTensor' only dense CPU tensors can be pinned
```

**To Reproduce**
```
ts = TimeSeries.from_series(series.dropna().astype('float32'), freq='D')

model = NBEATSModel(
    input_chunk_length=365, 
    output_chunk_length=51,
    torch_device_str='cuda',
    force_reset=True
)

model.fit(ts, verbose=True)
```
**System (please complete the following information):**
 - Python version: 3.7.7
 - darts version: 0.10.0

**Comment by novinsh** [comment]
Same problem here, I'm trying to fit the RNN model on GPU and this is the error that I get:

```RuntimeError: cannot pin 'torch.cuda.DoubleTensor' only dense CPU tensors can be pinned```

I also ran the example from [here](https://unit8co.github.io/darts/examples/05-RNN-examples.html) and the same error. The fit method runs without problem on the CPU but not on GPU.

Here's the error's backtrace: https://paste.ubuntu.com/p/KkpthPvPKp/

It seems that this problem has been introduced in the latest version (0.10.0), and it is not an issue in version 0.9.1 in which my code was written based on previously.

https://unit8co.github.io/darts/examples/05-RNN-examples.html

System:
- Python 3.8.5
- darts version: 0.10.0

**Comment by novinsh** [comment]
I tried by setting the pin_memory argument to False for [train_loader](https://github.com/unit8co/darts/blob/36ffcdbbd06c9aa927d0603c3481fdbaa2f04394/darts/models/torch_forecasting_model.py#L499) and [validation_loader](https://github.com/unit8co/darts/blob/36ffcdbbd06c9aa927d0603c3481fdbaa2f04394/darts/models/torch_forecasting_model.py#L508) and it worked like a charm. 

According to [this post](https://discuss.pytorch.org/t/when-to-set-pin-memory-to-true/19723) it appears that the data is already being loaded to the GPU before being passed to the data loader. I tried to trace it down but couldn't find any place which this would happen before the lines that I mentioned above. In the previous version (0.9.1) the dataset is made from TimeSeriesTorchDataset where the device is also set and it inherits from the Torch Dataset which is very clear and straight forward but now it is a bit ambiguous but it feels that something must be fixed in the `self._build_train_dataset`!

**Comment by hrzn** [comment]
Thank you for reporting, we'll fix this one asap.

**Comment by browserliu** [comment]
nbeats_model = NBEATSModel(
    input_chunk_length=30,
    output_chunk_length=7,
    generic_architecture=True,
    num_stacks=10,
    num_blocks=1,
    num_layers=4,
    layer_widths=512,
    n_epochs=10,
    nr_epochs_val_period=1,
    batch_size=800,
)

average_error = nbeats_model.backtest(srs, start=pd.Timestamp('2021-01-07 00:00:00'), forecast_horizon=168, verbose=False)



---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/tmp/ipykernel_102/4049927750.py in <module>
     12 )
     13 
---> 14 average_error = nbeats_model.backtest(srs, start=pd.Timestamp('2021-01-07 00:00:00'), forecast_horizon=168, verbose=False)

~/.local/lib/python3.7/site-packages/darts/models/forecasting_model.py in backtest(self, series, past_covariates, future_covariates, num_samples, start, forecast_horizon, stride, retrain, overlap_end, last_points_only, metric, reduction, verbose)
    421                                               overlap_end=overlap_end,
    422                                               last_points_only=last_points_only,
--> 423                                               verbose=verbose)
    424 
    425         if last_points_only:

~/.local/lib/python3.7/site-packages/darts/utils/utils.py in sanitized_method(self, *args, **kwargs)
    151 
    152                 getattr(self, sanity_check_method)(*only_args.values(), **only_kwargs)
--> 153             return method_to_sanitize(self, *only_args.values(), **only_kwargs)
    154         return sanitized_method
    155     return decorator

~/.local/lib/python3.7/site-packages/darts/models/forecasting_model.py in historical_forecasts(self, series, past_covariates, future_covariates, num_samples, start, forecast_horizon, stride, retrain, overlap_end, last_points_only, verbose)
    304 
    305             if retrain:
--> 306                 self._fit_wrapper(series=train, past_covariates=past_covariates, future_covariates=future_covariates)
    307 
    308             forecast = self._predict_wrapper(n=forecast_horizon,

~/.local/lib/python3.7/site-packages/darts/models/forecasting_model.py in _fit_wrapper(self, series, past_covariates, future_covariates)
    773     def _fit_wrapper(self, series: TimeSeries, past_covariates: Optional[TimeSeries],
    774                      future_covariates: Optional[TimeSeries]):
--> 775         self.fit(series, past_covariates=past_covariates, future_covariates=future_covariates)
    776 
    777 

~/.local/lib/python3.7/site-packages/darts/utils/torch.py in decorator(self, *args, **kwargs)
     63         with fork_rng():
     64             manual_seed(self._random_instance.randint(0, high=MAX_TORCH_SEED_VALUE))
---> 65             return decorated(self, *args, **kwargs)
     66     return decorator

~/.local/lib/python3.7/site-packages/darts/models/torch_forecasting_model.py in fit(self, series, past_covariates, future_covariates, val_series, val_past_covariates, val_future_covariates, verbose, epochs)
    429         logger.info('Train dataset contains {} samples.'.format(len(train_dataset)))
    430 
--> 431         self.fit_from_dataset(train_dataset, val_dataset, verbose, epochs)
    432 
    433     @random_method

~/.local/lib/python3.7/site-packages/darts/utils/torch.py in decorator(self, *args, **kwargs)
     63         with fork_rng():
     64             manual_seed(self._random_instance.randint(0, high=MAX_TORCH_SEED_VALUE))
---> 65             return decorated(self, *args, **kwargs)
     66     return decorator

~/.local/lib/python3.7/site-packages/darts/models/torch_forecasting_model.py in fit_from_dataset(self, train_dataset, val_dataset, verbose, epochs)
    517 
    518         # Train model
--> 519         self._train(train_loader, val_loader, tb_writer, verbose, train_num_epochs)
    520 
    521         # Close tensorboard writer

~/.local/lib/python3.7/site-packages/darts/models/torch_forecasting_model.py in _train(self, train_loader, val_loader, tb_writer, verbose, epochs)
    790             total_loss = 0
    791 
--> 792             for batch_idx, train_batch in enumerate(train_loader):
    793                 self.model.train()
    794                 output = self._produce_train_output(train_batch[:-1])

~/.local/lib/python3.7/site-packages/torch/utils/data/dataloader.py in __next__(self)
    515             if self._sampler_iter is None:
    516                 self._reset()
--> 517             data = self._next_data()
    518             self._num_yielded += 1
    519             if self._dataset_kind == _DatasetKind.Iterable and \

~/.local/lib/python3.7/site-packages/torch/utils/data/dataloader.py in _next_data(self)
    557         data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
    558         if self._pin_memory:
--> 559             data = _utils.pin_memory.pin_memory(data)
    560         return data
    561 

~/.local/lib/python3.7/site-packages/torch/utils/data/_utils/pin_memory.py in pin_memory(data)
     53         return type(data)(*(pin_memory(sample) for sample in data))
     54     elif isinstance(data, container_abcs.Sequence):
---> 55         return [pin_memory(sample) for sample in data]
     56     elif hasattr(data, "pin_memory"):
     57         return data.pin_memory()

~/.local/lib/python3.7/site-packages/torch/utils/data/_utils/pin_memory.py in <listcomp>(.0)
     53         return type(data)(*(pin_memory(sample) for sample in data))
     54     elif isinstance(data, container_abcs.Sequence):
---> 55         return [pin_memory(sample) for sample in data]
     56     elif hasattr(data, "pin_memory"):
     57         return data.pin_memory()

~/.local/lib/python3.7/site-packages/torch/utils/data/_utils/pin_memory.py in pin_memory(data)
     45 def pin_memory(data):
     46     if isinstance(data, torch.Tensor):
---> 47         return data.pin_memory()
     48     elif isinstance(data, string_classes):
     49         return data

RuntimeError: cannot pin 'torch.cuda.DoubleTensor' only dense CPU tensors can be pinned

-----------------------  system
torch.__version__ 1.8.1+cu102
3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37)  [GCC 9.3.0]
darts                             0.10.0

**Comment by hrzn** [comment]
The bug has been solved in v0.10.1. Thanks again for reporting.

**Comment by Ayesha-An** [comment]
0%|          | 0/28 [00:00<?, ?it/s]
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [18], in <cell line: 2>()
      1 model = create_model(params)
----> 2 model = train_and_validate(model, train_data, test_data, params)

Input In [16], in train_and_validate(model, train_dataset, val_dataset, params)
     17 optimizer = torch.optim.Adam(model.parameters(), lr=params["lr"])
     18 for epoch in range(1, params["epochs"] + 1):
---> 19     train(train_loader, model, criterion, optimizer, epoch, params)
     20     validate(val_loader, model, criterion, epoch, params)
     21 return model

Input In [13], in train(train_loader, model, criterion, optimizer, epoch, params)
      3 model.train()
      4 stream = tqdm(train_loader)
----> 5 for i, (images, target) in enumerate(stream, start=1):
      6     images = images.to(params["device"], non_blocking=True)
      7     target = target.to(params["device"], non_blocking=True)

File ~\Anaconda3\envs\py38\lib\site-packages\tqdm\std.py:1195, in tqdm.__iter__(self)
   1192 time = self._time
   1194 try:
-> 1195     for obj in iterable:
   1196         yield obj
   1197         # Update and possibly print the progressbar.
   1198         # Note: does not call self.update(1) for speed optimisation.

File ~\Anaconda3\envs\py38\lib\site-packages\torch\utils\data\dataloader.py:681, in _BaseDataLoaderIter.__next__(self)
    678 if self._sampler_iter is None:
    679     # TODO(https://github.com/pytorch/pytorch/issues/76750)
    680     self._reset()  # type: ignore[call-arg]
--> 681 data = self._next_data()
    682 self._num_yielded += 1
    683 if self._dataset_kind == _DatasetKind.Iterable and \
    684         self._IterableDataset_len_called is not None and \
    685         self._num_yielded > self._IterableDataset_len_called:

File ~\Anaconda3\envs\py38\lib\site-packages\torch\utils\data\dataloader.py:723, in _SingleProcessDataLoaderIter._next_data(self)
    721 data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
    722 if self._pin_memory:
--> 723     data = _utils.pin_memory.pin_memory(data, self._pin_memory_device)
    724 return data

File ~\Anaconda3\envs\py38\lib\site-packages\torch\utils\data\_utils\pin_memory.py:65, in pin_memory(data, device)
     63 elif isinstance(data, collections.abc.Sequence):
     64     try:
---> 65         return type(data)([pin_memory(sample, device) for sample in data])  # type: ignore[call-arg]
     66     except TypeError:
     67         # The sequence type may not support `__init__(iterable)` (e.g., `range`).
     68         return [pin_memory(sample, device) for sample in data]

File ~\Anaconda3\envs\py38\lib\site-packages\torch\utils\data\_utils\pin_memory.py:65, in <listcomp>(.0)
     63 elif isinstance(data, collections.abc.Sequence):
     64     try:
---> 65         return type(data)([pin_memory(sample, device) for sample in data])  # type: ignore[call-arg]
     66     except TypeError:
     67         # The sequence type may not support `__init__(iterable)` (e.g., `range`).
     68         return [pin_memory(sample, device) for sample in data]

File ~\Anaconda3\envs\py38\lib\site-packages\torch\utils\data\_utils\pin_memory.py:50, in pin_memory(data, device)
     48 def pin_memory(data, device=None):
     49     if isinstance(data, torch.Tensor):
---> 50         return data.pin_memory(device)
     51     elif isinstance(data, string_classes):
     52         return data

RuntimeError: cannot pin 'torch.cuda.FloatTensor' only dense CPU tensors can be pinned



python 3.8
